File: frontend/src/stores/recipes.js
import { defineStore } from 'pinia'
import api from '@/api'

export const useRecipesStore = defineStore('recipes', {
  state: () => ({
    recipes: [],
    currentRecipe: null,
    total: 0,
    page: 1,
    pageSize: 12,
    loading: false,
    error: null,
    filters: {
      category: null,
      tag: null,
      search: '',
      sort: 'newest'
    }
  }),

  actions: {
    async fetchRecipes(reset = false) {
      if (reset) {
        this.page = 1
        this.recipes = []
      }
      this.loading = true
      this.error = null
      try {
        const params = {
          page: this.page,
          page_size: this.pageSize,
          sort: this.filters.sort
        }
        if (this.filters.category) params.category = this.filters.category
        if (this.filters.tag) params.tag = this.filters.tag
        if (this.filters.search) params.search = this.filters.search

        const result = await api.getRecipes(params)
        this.recipes = result.data || []
        this.total = result.total || 0
      } catch (e) {
        this.error = e.message
      } finally {
        this.loading = false
      }
    },

    async fetchRecipe(slug) {
      this.loading = true
      this.error = null
      try {
        this.currentRecipe = await api.getRecipe(slug)
      } catch (e) {
        this.error = e.message
        this.currentRecipe = null
      } finally {
        this.loading = false
      }
    },

    setFilter(key, value) {
      this.filters[key] = value
      this.fetchRecipes(true)
    },

    clearFilters() {
      this.filters = { category: null, tag: null, search: '', sort: 'newest' }
      this.fetchRecipes(true)
    }
  }
})

File: routes/recipes.py
"""Public recipe routes."""
from typing import Optional
from fastapi import APIRouter, HTTPException, Query

from models.schemas import DIETARY_TAGS
from utils.repository import db

router = APIRouter(prefix="/api", tags=["recipes"])


@router.get("/tags")
async def list_tags():
    return DIETARY_TAGS


@router.get("/recipes")
async def list_recipes(
    category: Optional[str] = Query(None, description="Filter by category slug"),
    tag: Optional[str] = Query(None, description="Filter by tag slug"),
    search: Optional[str] = Query(None, description="Search by name"),
    sort: Optional[str] = Query("newest", description="Sort: newest, stars, name"),
    page: int = Query(1, ge=1),
    page_size: int = Query(12, ge=1, le=50),
):
    # Resolve category slug to ID
    category_id = None
    if category:
        cat = db.get_by_field("categories", "slug", category)
        if cat:
            category_id = cat["id"]

    offset = (page - 1) * page_size
    recipes, total = db.query_recipes(
        category_id=category_id,
        tag=tag,
        search=search,
        sort_by=sort,
        offset=offset,
        limit=page_size,
    )

    return {"data": recipes, "total": total, "page": page, "page_size": page_size}


@router.get("/recipes/{slug}")
async def get_recipe(slug: str):
    recipe = db.get_by_field("recipes", "slug", slug)
    if not recipe:
        raise HTTPException(status_code=404, detail="Recipe not found")
    return recipe
